### ISSUE TITLE
Implements debouncing in mathjax

### ISSUE THREAD

**Issue by Arthur-Milchior** [ankitects/anki#3803]
When I'm writting mathjax, the editor freeze entirely, which makes it hard to use. I think it should be debounced. That is, when there is some change in the mathjax editor, wait, let's say, one second without any changes before sending it to be displayed. 

This would ensure that at most one rendering request is done per second, instead of asking to render all intermediate steps, most of them not being proper mathjax.

Even if I can blind type and do relatively few typos, having the screen frozen sometime for 10 seconds is frustrating simply because when I type an expression such as
```
[F(\alpha_1,\dots,\alpha_g,\alpha_{g+1}):F]= [F(\alpha_1,\dots,\alpha_g,\alpha_{g+1}):F(\alpha_1,\dots,\alpha_g)][F(\alpha_1,\dots,\alpha_g):F]= [(F(\alpha_1,\dots,\alpha_g))(\alpha_{g+1}):F(\alpha_1,\dots,\alpha_g)][F(\alpha_1,\dots,\alpha_g):F] \le [F(\alpha_{g+1}):F]\prod_{a=1}^g\deg\alpha_a=\alpha_{g+1}\prod_{a=1}^g\deg\alpha_i=\prod_{a=1}^{g+1}\deg\alpha_i
```

I want to do copy/paste a lot instead of retyping everything. And I can't copy anything when the UI is frozen.

**Comment by rlanday** [comment]
I started going through the editor code a while ago because I was experiencing all sorts of bugs, particularly involving flashcards with excerpts from machine learning papers, where there would be some MathJax equations interleaved with SVG files. I used to work at Google fixing text input and editing bugs on Chrome for Android, so I thought it wouldn’t be too hard to fix some of the Anki bugs I’m running into.

Unfortunately I what I found is that the current editor is essentially a series of hacks built on top of other hacks, I believe without any test cases to prevent regressions; you’re essentially proposing to add yet another hack, which is not really going to fix the underlying problem with MathJax editing, but is going to make the code base even more difficult to understand and make changes to. The current editor is built on contenteditable and execCommand, which is an approach that the browser and web standards people strongly believe is never ever going to work properly. See the below email from the W3C public-editing-tf mailing list:

https://lists.w3.org/Archives/Public/public-editing-tf/2024Sep/0000.html
> **ExecCommand IS NOT a viable way of creating a richtext editor** for almost
any usecase and hasn't been for the past 10-15 years. Many years of
developer lives have been wasted attempting to do so, which is why **we
added a huge warning to the beginning of the document** trying to ensure
that no one else wastes times trying to do that.

> We have spent the past decade creating the alternative you are asking for.
Check out EditContext and Input Events. Input Events is already shipped in
all major browsers.

> However, these provide quite foundational APIs and a lot of work has to be
done in JavaScript to create a viable editor. For most users who need a
custom JavaScript editor, I would personally recommend using one of the
existing editor libraries (like CKEditor 5 or ProseMirror)  and build their
editor on top of that.

The correct solution here is almost certainly to integrate one of these third-party editors into Anki. There’s already an add-on that adds a button to edit a card with TinyMCE:
https://ankiweb.net/shared/info/805891399

I tried it but this is not a solution to my problems since I don’t think this editor lets you add SVG files. But I assume this is a solvable problem, or maybe one of the other editors would be more suitable. The TinyMCE add-on avoids the MathJax hangs by not attempting to render equations inside the editor, which is maybe preferable to having to encountering a hang and having to force quit Anki resulting in data loss, but is also not as nice as rendering equations inline.

**Comment by Arthur-Milchior** [comment]
Thank you so much @iamllama I just tested it and it seems to work very well, you really improved my editing of mathematics notes!!!!

@rlanday Hey, fellow Chromie!

I'm happy you're wrong about the complexity of the change I requested.

I guess you're right at least in the sense that execCommand may be removed which will then break anki and any mobile version reusing this editor.
However, I doubt this is the right place for this discussion, given that you're requesting a fundamental architectural change which seems unrelated to the small issue I raised.

**Comment by rlanday** [comment]
This specific change might work as a Band-Aid fix; I’ll leave that up to the editing maintainers to figure out. I’ll note once again that there doesn’t appear to be an associated test case for this behavior, so I don’t know how you’ll make sure this doesn’t break some other bug that was already fixed, or some other change won’t break this again.

If I open a separate bug for integrating a third-party editor, is that a discussion people would be open to having? I have a whole series of bugs I could file with repro steps about redo/undo being flaky, copy and paste sometimes being broken, etc. The most serious is that the editor sometimes completely locks up, I think because it runs some JavaScript (relating to the selection handles?) that gets caught in an infinite loop.

**Comment by rlanday** [comment]
I just installed Anki 25.07 and am actually a bit disappointed by this new behavior. If I’m quickly typing MathJax, it’s not that it renders at most twice a second; it actually doesn’t render at all until half a second after I stop typing. This makes it a lot more difficult to tell if/where I’ve introduced a syntax error.

I’m skeptical that this is really the correct fix. For OP’s issue with copy/paste, I think it would be perfectly reasonable to not render anything until after the paste operation is finished; but I suspect this is the default behavior anyway assuming the paste option is atomic? If we want to debounce while typing, the correct fix would be to wait to render until 500 ms after the last render attempt, not 500 ms after the user stops typing (I assume rendering twice per second is acceptable but please correct me if that itself would pose a problem).

Edit: I think I misunderstood; the reported issue seems to be that the UI freezes during typing, which makes copy/paste impossible at that time.
